### PR TITLE
Implement Jörmungandr TransactionLayer

### PR DESCRIPTION
**Pull request by Anviking**
# Issue Number

#219 

~~:warning: base branch is `KtorZ/transaction-id-golden-tests` :warning:~~

# Overview

- [x] I have implemented a preliminary `TransactionLayer`

# Comments

- [x] Golden tests are wip (not pushed). ~I'd call for merging this as soon as possible still.~

### CODE REVIEW

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 1]
```diff
@@ -1,19 +1,67 @@
...
+    ( ByteString )
 
+import qualified Cardano.Crypto.Wallet as CC
 
 -- | Construct a 'TransactionLayer' compatible with Shelley and 'Jörmungandr'
-newTransactionLayer :: TransactionLayer t
-newTransactionLayer = TransactionLayer
-    { mkStdTx = error "TODO: See http-bridge as starting point"
+newTransactionLayer
+    :: forall n . Hash "Block0Hash"
```
Comment (Anviking) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: @KtorZ 

> I am not sure about this type ... And, does this really have to be the hash from the first block header?

To my knowledge — yes

**Comment by Anviking** [comment]
I resurrected this from #418 and merged in the `\-> Quantity 0` from #437.

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs (hunk 1 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/witness.rs#L60-L62

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 2]
```diff
@@ -1,19 +1,73 @@
...
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
 
 module Cardano.Wallet.Jormungandr.Transaction
     ( newTransactionLayer
     ) where
 
 import Prelude
 
+import Cardano.Wallet.Jormungandr.Compatibility
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: let's be more specific here : 
```
import Cardano.Wallet.Jormungandr.Compatibility
    ( Jormungandr )
```

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs [hunk 3]
```diff
@@ -1,19 +1,73 @@
...
+  where
+    withEither :: e -> Maybe a -> Either e a
+    withEither e = maybe (Left e) Right
+
+newtype WitnessData = WitnessData ByteString
+
+-- Rust implementation:
+--
+-- https://github.com/input-output-hk/rust-cardano/blob/13af7aa08cdac8efc9dc8b0f7ff9162d28b2d479/chain-impl-mockchain/src/transaction/witness.rs#L60-L61
+witnessUtxoData :: Hash "Block0Hash" -> Hash "Tx" -> WitnessData
```
Comment (paweljakubas) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: minor stuff, but maybe we would like to have all functions below in `where` of newTransactionLayer` ???

**Comment by Anviking** [comment]
Hm… the diff is messed up since I force-pushed `anviking/219/txid` :/

**Comment by paweljakubas** [comment]
general comment : the signing looks good at first sight, but I will be persuaded if extensive golden tests will support this feeling :-)

**Review by paweljakubas — APPROVED** [review]

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs (hunk 3 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs]: I was thinking one _might_ want to export and test individually. And `witnessUtxoData` will be (one of several) "constructors" for `WitnessData`, so I think they should be on the same level as the `WitnessData` declaration.

Happy to surrender though. 

cc @KtorZ 

> Why the WitnessData indirection here? Why not have the hashes as arguments directly?

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 4]
```diff
@@ -203,21 +212,26 @@ getMessage = label "getMessage" $ do
         5 -> unimpl
         other -> fail $ "Unexpected content type tag " ++ show other
 
+putMessage :: Message -> Put
+putMessage m = withSizeHeader16be $ case m of
+    Transaction (tx, wits) -> do
+        putWord8 2
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: we should probably add a comment explaining why constant 2 (or reference to the actuall rust api line containing this info)

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 5]
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
 putWitness :: TxWitness -> Put
 putWitness witness =
     case witness of
-        PublicKeyWitness xPub (Hash sig) -> do
+        PublicKeyWitness _xPub (Hash sig) -> do
             -- Witness sum type:
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: we might want to define sum type that would describe the constants in this module, at least for documentation

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 6]
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
...
-            --   * 1 for old address witnness scheme
-            --   * 2 for new address witness scheme
-            --   * 3 for account witness
+            --   * 0 for old address witnness scheme
+            --   * 1 for new address witness scheme
+            --   * 2 for account witness
+            --   * 3 for multisig
             putWord8 1
-            putByteString xPub
+            --putByteString xPub
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: But then old address scheme won't work, right? (as it expects xPub in its format)

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs [hunk 7]
```diff
@@ -244,37 +258,43 @@ putSignedTransaction (tx, witnesses) = do
...
+putTxBody :: Tx -> Put
+putTxBody (Tx inputs outputs) = do
     mapM_ putInput inputs
     mapM_ putOutput outputs
   where
     putInput (TxIn inputId inputIx) = do
         -- NOTE: special value 0xff indicates account spending
         -- only old utxo/address scheme supported for now
         putWord8 $ fromIntegral inputIx
+        putWord64be 14 -- TODO: We need input value!
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Binary.hs]: I am not sure I understand the idea behind this ?

EDIT: aha - I see https://github.com/input-output-hk/rust-cardano/blob/13af7aa08cdac8efc9dc8b0f7ff9162d28b2d479/chain-impl-mockchain/src/transaction/transfer.rs#L21 :+1:

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Compatibility.hs [hunk 8]
```diff
@@ -84,7 +79,7 @@ genesis = BlockHeader
 -- The corresponding rust implementation is:
 -- https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/transaction.rs#L115-L119
 instance TxId (Jormungandr n) where
-    txId = blake2b256 . putTokenTransfer
+    txId = blake2b256 . putTxBody
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Compatibility.hs]: :+1: https://github.com/input-output-hk/rust-cardano/blob/e5d974f7bedeb00c9c9d688ac66094a34bf8f40d/chain-impl-mockchain/src/transaction/transaction.rs#L117

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs [hunk 9]
```diff
@@ -185,6 +199,20 @@ mkJormungandrLayer mgr baseUrl = JormungandrLayer
                         (BlockId parentId)
                         (Just count)
                 left ErrGetDescendantsNetworkUnreachable <$> defaultHandler ctx x
+    , postSignedTx = \tx -> ExceptT $ do
+        let e0 = "?"
```
Comment (akegalj) [inline lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs]: I can't find where this was taken from?

File: lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs [hunk 10]
```diff
@@ -171,3 +211,150 @@ instance Arbitrary (Hash "Tx") where
...
+            let (key, pwd) = unsafeKey $ fromHex
+                    "4086a5c60a633a5d26ce619ceac000be4240c8187c15ef51ea59220c05\
+                    \08184fc11cd11686252eefbccc20eb3d6743f2f0cb6842baf939e7d91c\
+                    \96963ebe5eac054bba3fcb7b5f2491861b6e2b8bf0a3f9257cdf2fdf41\
+                    \ca5af7072e77c70f3a578f464364f1442b79b05e92f3df7ac4c1f7f891\
+                    \03c1de1e92c17247a77e7f18"
+            let keyFrom _ = Just (key, pwd)
+            let inps = pure
+                    ( txin
+                    , TxOut (Address "used for lookup") (Coin 14)
```
Comment (akegalj) [inline lib/jormungandr/test/unit/Cardano/Wallet/Jormungandr/TransactionSpec.hs]: address content is not important?

**Review by akegalj — COMMENTED** [review]
It mostly looks good to me. There are few places where I am not sure in the correctness (I wasn't able to find the corresponding place in rust code)

File: lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs (hunk 9 above)
Comment (Anviking) [reply lib/jormungandr/src/Cardano/Wallet/Jormungandr/Network.hs]: Sorry, for the unnecessary cognitive load by including this commit. (I shouldn't have!) I copied it from `http-bridge` (for testing purposes) where `?` was an error message. I have not at all looked into what kind of errors jormungandr returns. Dropping it now.

**Comment by KtorZ** [comment]
This PR is to merge after:

- #451 
- #452 

- [x] Added golden tests for transactions signing using `jcli` (only new address format for now though..)
- [x] Removed commented-code or unused code
- [x] Did some refactoring, replaced custom functions with built-in, removed unneeded indirections and renamed some functions / top-level expressions
- [x] Fixed bug in the transaction signer for Jormungandr, it used to use the transaction id as a payload, whereas it should be using the id from each transaction input instead!

cc @paweljakubas @Anviking

**Pull request merged**